### PR TITLE
Flat dependency trees

### PR DESCRIPTION
**Pull request by asafgabai**
- [x] All [tests](https://github.com/jfrog/ide-plugins-common/actions/workflows/test.yml) passed. If this feature is not already covered by the tests, I added new tests.
-----
A new structure for the dependency trees: created a new class for dependency trees (DepTree) that has a flat tree structure to save memory.

Depends on https://github.com/jfrog/gradle-dep-tree/pull/14

### CODE REVIEW

File: settings.gradle [hunk 1]
```diff
@@ -1 +1,7 @@
+sourceControl {
```
Comment (asafgabai) [inline settings.gradle:1]: Reminder for myself to remove this when it's released.

File: src/main/java/com/jfrog/ide/common/scan/GraphScanLogic.java [hunk 2]
```diff
@@ -98,25 +100,35 @@ DependencyTree createScanTree(DependencyTree root) {
...
-            if (((DependencyTree) child.getParent()).isMetadata()) {
-                // All direct dependencies should be in the cache. This line make sure that dependencies that
-                // wouldn't return from Xray will not be scanned again during the next quick scan.
-                String componentId = contains(childFullId, "://") ?
-                        substringAfter(childFullId, "://") : childFullId;
-            }
-            if (componentsAdded.add(child.getComponentId())) {
-                scanTree.add(new DependencyTree(child.getComponentId()));
+            if (!child.isMetadata()) {
+                // Add the dependency subtree to the scan tree
```
Comment (talarian1) [inline src/main/java/com/jfrog/ide/common/scan/GraphScanLogic.java:104]:
```suggestion
                // Add the dependency subtree to the scan tree if it hasn't been added yet. 
```

File: src/main/java/com/jfrog/ide/common/scan/GraphScanLogic.java [hunk 3]
```diff
@@ -98,25 +100,35 @@ DependencyTree createScanTree(DependencyTree root) {
...
+                if (componentsAdded.add(child.getComponentId())) {
+                    scanTree.add(new DependencyTree(child.getComponentId()));
+                }
             }
             populateScanTree(child, scanTree, componentsAdded);
```
Comment (talarian1) [inline src/main/java/com/jfrog/ide/common/scan/GraphScanLogic.java:109]: Should we populateScanTree if it isn't the first time we "see" this dependency?
`(componentsAdded.add(child.getComponentId() == false)`

File: src/main/java/com/jfrog/ide/common/yarn/YarnTreeBuilder.java [hunk 4]
```diff
@@ -42,115 +39,104 @@ public YarnTreeBuilder(Path projectDir, Map<String, String> env) {
...
+        String packageName;
         JsonNode nameNode = packageJson.get("name");
         if (nameNode != null) {
-            return nameNode.asText() + getPostfix(logger, npmLsResults);
-        }
-        if (projectDir.getFileName() != null) {
-            return projectDir.getFileName().getFileName().toString() + getPostfix(logger, npmLsResults);
+            packageName = nameNode.asText();
+        } else if (projectDir.getFileName() != null) {
+            packageName = projectDir.getFileName().getFileName().toString();
```
Comment (talarian1) [inline src/main/java/com/jfrog/ide/common/yarn/YarnTreeBuilder.java:123]: I'm almost sure that the second getFileName() call will do nothing :)

File: src/main/java/com/jfrog/ide/common/deptree/DepTree.java [hunk 5]
```diff
@@ -0,0 +1,26 @@
+package com.jfrog.ide.common.deptree;
+
+import java.util.Map;
+
+public class DepTree {
```
Comment (talarian1) [inline src/main/java/com/jfrog/ide/common/deptree/DepTree.java:5]: It is a bit confusing that we have both DepTree and DependencyTree classes...
What do you think?

**Review by talarian1 — CHANGES_REQUESTED** [review]

File: src/main/java/com/jfrog/ide/common/go/GoTreeBuilder.java [hunk 6]
```diff
@@ -34,39 +35,63 @@ public class GoTreeBuilder {
...
-            Version goVersion = parseGoVersion(versionRes, logger);
-            goDriver.modTidy(false, goVersion.isAtLeast(MIN_GO_VERSION));
-            DependencyTree rootNode = GoDependencyTree.createDependencyTree(goDriver, logger, false, goVersion.isAtLeast(MIN_GO_VERSION_FOR_BUILD_VCS_FLAG));
-            addGoVersionNode(rootNode, goVersion);
-            setGeneralInfo(rootNode);
-            setNoneScope(rootNode);
-            return rootNode;
-        } finally {
-            FileUtils.deleteDirectory(tmpDir);
+        // Run go list -f "{{with .Module}}{{.Path}} {{.Version}}{{end}}" all
```
[the remark this answers is not in the document]
Comment (asafgabai) [reply src/main/java/com/jfrog/ide/common/go/GoTreeBuilder.java:65]: I think you're right, and actually, it's already there, so I'll remove it from here.

File: src/main/java/com/jfrog/ide/common/gradle/GradleTreeBuilder.java [hunk 7]
```diff
@@ -59,63 +55,30 @@ public DependencyTree buildTree(Log logger) throws IOException {
...
-            rootNode.add(projectNode);
+            GradleDepTreeResults results = objectMapper.readValue(projectFile, GradleDepTreeResults.class);
+            for (Map.Entry<String, GradleDependencyNode> nodeEntry : results.getNodes().entrySet()) {
+                String compId = nodeEntry.getKey();
+                GradleDependencyNode gradleDep = nodeEntry.getValue();
+                DepTreeNode node = new DepTreeNode().scopes(gradleDep.getConfigurations()).children(gradleDep.getChildren());
+                nodes.put(compId, node);
+            }
+            nodes.get(results.getRoot()).descriptorFilePath(descriptorFilePath);
+            rootNode.getChildren().add(results.getRoot());
```
[the remark this answers is not in the document]
Comment (asafgabai) [reply src/main/java/com/jfrog/ide/common/gradle/GradleTreeBuilder.java:72]: `nodes.get(results.getRoot())` would give us the root `DepTreeNode` object, but we need the root ID, which is `results.getRoot()`.
I did a little change there to make it clearer.

File: src/main/java/com/jfrog/ide/common/gradle/GradleTreeBuilder.java [hunk 8]
```diff
@@ -59,63 +55,30 @@ public DependencyTree buildTree(Log logger) throws IOException {
      * @return a dependency tree contain one or more Gradle projects.
      * @throws IOException in case of any I/O error.
      */
-    private DependencyTree createDependencyTrees(List<File> gradleDependenciesFiles) throws IOException {
-        DependencyTree rootNode = new DependencyTree(projectDir.getFileName().toString());
-        rootNode.setMetadata(true);
-        rootNode.setGeneralInfo(new GeneralInfo().componentId(projectDir.getFileName().toString()).path(projectDir.toString()));
+    private DepTree createDependencyTrees(List<File> gradleDependenciesFiles) throws IOException {
+        String rootId = projectDir.getFileName().toString();
+        DepTreeNode rootNode = new DepTreeNode().descriptorFilePath(descriptorFilePath);
```
[the remark this answers is not in the document]
Comment (asafgabai) [reply src/main/java/com/jfrog/ide/common/gradle/GradleTreeBuilder.java:60]: Nope. The `metadata` field wasn't needed anymore and was removed 🕺

File: src/main/java/com/jfrog/ide/common/scan/GraphScanLogic.java (hunk 2 above)
Comment (asafgabai) [reply src/main/java/com/jfrog/ide/common/scan/GraphScanLogic.java:104]: Actually, this method was replaced by another one and is not used anymore. I'll remove it.

**Pull request merged**